### DISCUSSION TITLE
Add params to Panel Lifecycle hooks

### DISCUSSION THREAD

**Discussion by dbpolito** [filamentphp/filament#8179] [Ideas]
https://filamentphp.com/docs/3.x/panels/resources/creating-records#lifecycle-hooks

Currently no params are passed to `beforeCreate`, `afterCreate`, etc... i think it would be useful add available data to these events to allow more customization, like:

```
beforeCreate(array $data)

afterCreate(Model $record, array $data)
```

Wanted to validate that first before i open a PR.

**Comment by dbpolito** [comment]
Also, good to mention that introducing these params wouldn't create breaking changes...

**Comment by danharrin** [comment]
You can access `$this->record` and `$this->data`

**Reply by dbpolito** [reply]
true for `$this->record`, but for `$this->data` i don't see where it's used, only here https://github.com/filamentphp/filament/blob/3.x/packages/panels/src/Resources/Pages/EditRecord.php#L98-L101 which seems to not be used at all...

we could use `$this->form->getState()` but we would lose the changes on `mutateFormDataBeforeSave`:

https://github.com/filamentphp/filament/blob/3.x/packages/panels/src/Resources/Pages/EditRecord.php#L139

Anyway... still feels a nice addition...

**Reply by danharrin** [reply]
What are you trying to achieve? Maybe override `handleRecordCreation()` and call `parent::handleRecordCreation()`, which gives you access to the mutated form data

**Reply by dbpolito** [reply]
yeah, that's what i'm using... i wanted to create some traits to handle general pages, that could benefit having a more robust event system to improve, and thought it would be a good addition to the core as others might also benefit from it.

Anyway, it's not blocking me or anything, wanted to suggest improvements to the core 👍
